src/System.IO.Pipelines/System/IO/Pipelines/ReadCursor.cs

```diff
@@ -122,15 +156,18 @@ private ReadCursor SeekMultiSegment(long bytes, ReadCursor end, bool checkEndRea
             ReadCursor result = default;
             bool foundResult = false;
 
-            foreach (var segmentPart in new SegmentEnumerator(this, end))
+            var enumerator = new BufferEnumerator(this, end);
+            Memory<byte> memory;
+            while (enumerator.MoveNext())
             {
+                memory = enumerator.Current;
                 // We need to loop up until the end to make sure start and end are connected
                 // if end is not trusted
                 if (!foundResult)
                 {
-                    if (segmentPart.Length >= bytes)
+                    if (memory.Length >= bytes)
                     {
-                        result = new ReadCursor(segmentPart.Segment, segmentPart.Start + (int)bytes);
+                        result = enumerator.CreateCursor((int)bytes);
                         foundResult = true;
                         if (!checkEndReachable)
```

Comment:
I don't even like checkEndReachable as an option. For those willing to sacrifice perf for a runtime correctness check, I would just add a `bool ReadCursor.CheckEndReachable(ReadCursor end)`.

Reply:
As it stands, I think this never throws if `ReadCursor.IsDefault` since `Segment != end.Segment` and `enumerator.MoveNext()` would return false. I don't think we should even allow null Segments. It doesn't seem well supported to begin with.

Reply:
https://github.com/dotnet/corefxlab/issues/1969